python/xorbits/core/data.py

```diff
@@ -212,6 +212,23 @@ def gen():
 
                 yield from gen()
 
+    def __len__(self):
+        from .._mars.core import HasShapeTileable
+        from .execution import run
+
+        if isinstance(self.data._mars_entity, HasShapeTileable):
+            try:
+                return int(self.data._mars_entity.shape[0])
+            except (IndexError, ValueError):
+                # shape is unknown, execute it
+                run(self)
+                try:
+                    return int(self.data._mars_entity.shape[0])
+                except (IndexError, ValueError):  # happens when dimension is 0
+                    return 0
```

Comment:
The length of a zero-dimension object is quite confusing. In numpy, an error will be raised when you call len() on a zero-dimention array:

```python
In [5]: a = np.array(1)

In [6]: a.shape
Out[6]: ()

In [7]: len(a)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 1
----> 1 len(a)

TypeError: len() of unsized object
```

Would it be better if we just throw the IndexError or ValueError here?

Reply:
We can catch the IndexError and raise TypeError.